File: benchmarks/nyc_taxi/python/q11_heavy_summarize.py
import argparse
import time

import pyarrow.dataset as ds


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--dataset", required=True, help="Path to the partitioned parquet dataset")
    args = parser.parse_args()

    started = time.perf_counter()
    dataset = ds.dataset(args.dataset, format="parquet", partitioning="hive")

    table = dataset.to_table(
        columns=[
            "VendorID",
            "fare_amount",
            "tip_amount",
            "total_amount",
            "trip_distance",
            "passenger_count",
            "PULocationID",
        ]
    )
    frame = table.to_pandas()
    frame = frame[frame["trip_distance"].notna()]
    result = (
        frame.groupby("VendorID", dropna=False)
        .agg(
            mean_fare=("fare_amount", "mean"),
            mean_tip=("tip_amount", "mean"),
            total_rev=("total_amount", "sum"),
            max_dist=("trip_distance", "max"),
            min_dist=("trip_distance", "min"),
            total_passengers=("passenger_count", "sum"),
            trips=("VendorID", "size"),
            unique_locs=("PULocationID", "nunique"),
        )
        .reset_index()
        .sort_values("VendorID")
    )

    print(result.head(10))
    print("ROWS_SCANNED=NA")
    print(f"ROWS_RETURNED={len(result.index)}")
    print(f"ELAPSED_SEC={time.perf_counter() - started:.6f}")


if __name__ == "__main__":
    main()

File: benchmarks/nyc_taxi/r/q11_heavy_summarize.R
args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 1) {
  stop("usage: Rscript q11_heavy_summarize.R <parquet-dataset-path>", call. = FALSE)
}

suppressPackageStartupMessages({
  library(arrow)
  library(dplyr)
})

started <- Sys.time()
dataset_path <- args[[1]]
ds_tbl <- open_dataset(dataset_path, format = "parquet")

result <- ds_tbl |>
  filter(!is.na(trip_distance)) |>
  group_by(VendorID) |>
  summarise(
    mean_fare = mean(fare_amount, na.rm = TRUE),
    mean_tip = mean(tip_amount, na.rm = TRUE),
    total_rev = sum(total_amount, na.rm = TRUE),
    max_dist = max(trip_distance, na.rm = TRUE),
    min_dist = min(trip_distance, na.rm = TRUE),
    total_passengers = sum(passenger_count, na.rm = TRUE),
    trips = n(),
    unique_locs = n_distinct(PULocationID)
  ) |>
  collect()

print(head(result, 10))
cat("ROWS_SCANNED=NA\n")
cat(sprintf("ROWS_RETURNED=%s\n", nrow(result)))
cat(sprintf("ELAPSED_SEC=%.6f\n", as.numeric(difftime(Sys.time(), started, units = "secs"))))
